### PR TITLE
feat(languages): hcl

### PR DESCRIPTION
**Pull request by matoous**
Add support for HCL language.

The treesitter grammer used is https://github.com/MichaHoffmann/tree-sitter-hcl which is listed on official treesitter grammer page. There's also https://github.com/mitchellh/tree-sitter-hcl that has more stars 🤷 

- [ ] add also tf lsp

### CODE REVIEW

**Comment by matoous** [comment]
Queries are currently not working, need to learn how to write those.

File: runtime/queries/hcl/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  (template_directive_start) ; %{
+  (template_directive_end) ; }
+  (strip_marker) ; ~
+] @punctuation.special
+
+(numeric_lit) @number
+(bool_lit) @boolean
+(null_lit) @constant
+(comment) @comment
+(identifier) @variable
```
Comment (the-mikedavis) [inline runtime/queries/hcl/highlights.scm:82]: I think these are from the nvim-treesitter, yeah? nvim-treesitter's queries have backwards precedence (first rule top-to-bottom is considered last) while helix and tree-sitter-cli have it forwards (last rule top-to-bottom is considered last), so the identifier here is matching before the ones below on lines 93 and 94

if you move this 'default' rule for identifier below the other matches I think the queries will improve :+1:

Comment (matoous) [reply runtime/queries/hcl/highlights.scm]: Was fighting with it for some time but didn't manage to make it work, I ended up using https://github.com/mitchellh/tree-sitter-hcl that I was able to setup but it seems to have less highlighting.

**Comment by matoous** [comment]
Screen Shot 2022-01-13 at 21 36 16

This still seems off slightly. But slowly getting there.

File: runtime/queries/hcl/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+; (object_elem (identifier) () (expression (variable_expr (identifier) @field)))
+
+; (object_elem val: (expression
+;   (variable_expr
+;     (identifier) @type.builtin (#any-of? @type.builtin "bool" "string" "number" "object" "tuple" "list" "map" "set" "any"))))
+
+(get_attr (identifier) @variable.builtin (#any-of? @variable.builtin  "root" "cwd" "module"))
+((identifier) @keyword (#any-of? @keyword "module" "root" "cwd" "resource" "variable" "data" "locals" "terraform" "provider" "output"))
+((identifier) @type.builtin (#any-of? @type.builtin "bool" "string" "number" "object" "tuple" "list" "map" "set" "any"))
+(variable_expr (identifier) @variable.builtin (#any-of? @variable.builtin "var" "local" "path"))
```
Comment (the-mikedavis) [inline runtime/queries/hcl/highlights.scm:13]: helix doesn't have the `#any-of` predicate, so you'll need to convert these to an equivalent `#match?` in a format like

```tsq
(get_attr
  (identifier) @variable.builtin
  (#match? @variable.builtin "^(root|cwd|module)$"))
```

I think that should improve the highlights a bit :+1:

**Comment by the-mikedavis** [comment]
I thought this branch of mine was lost but I had it locally: [the-mikedavis/helix@`master...md-tree-sitter-hcl`#diff-f2a539ff2e](https://github.com/the-mikedavis/helix/compare/master...md-tree-sitter-hcl#diff-f2a539ff2e301e5bfafb182be519173222a3dd3c69bb0bf59be4fb2adff8e12d)

I was originally trying out the MichaHoffmann grammar and it seemed pretty reasonable. This mitchellh one seems good too though :monocle_face:

**Comment by archseer** [comment]
https://github.com/helix-editor/helix/pull/1705

**Pull request closed without merging**